janus_collector/src/lib.rs

```diff
@@ -0,0 +1,866 @@
+    pub async fn start_collection(
+        &self,
+        batch_interval: Interval,
+        aggregation_parameter: &V::AggregationParam,
+    ) -> Result<CollectJob<V>, Error> {
+        let collect_request = CollectReq {
+            task_id: self.parameters.task_id,
+            batch_interval,
+            agg_param: aggregation_parameter.get_encoded(),
+        };
+        let url = self.parameters.collect_endpoint()?;
+
+        let response = retry_http_request(
+            self.parameters.http_request_retry_parameters.clone(),
+            || async {
+                let mut request = self
+                    .http_client
+                    .post(url.clone())
+                    .header(CONTENT_TYPE, CollectReq::MEDIA_TYPE)
+                    .body(collect_request.get_encoded());
+                match &self.parameters.authentication {
+                    Authentication::BearerToken(token) => {
+                        request = request.header(DAP_AUTH_HEADER, token)
+                    }
+                }
+                request.send().await
+            },
+        )
+        .await
+        .map_err(|res| match res {
+            Ok(response) => Error::Http(response.status()),
+            Err(error) => Error::HttpClient(error),
+        })?;
+
+        let status = response.status();
+        if status != StatusCode::SEE_OTHER {
+            return Err(Error::Http(status));
+        }
+        let location_header_value = response
+            .headers()
+            .get(LOCATION)
+            .ok_or(Error::MissingLocationHeader)?
+            .to_str()?;
+        let collect_job_url = location_header_value.parse()?;
+
+        Ok(CollectJob {
+            collect_job_url,
+            batch_interval,
+            aggregation_parameter: aggregation_parameter.clone(),
+        })
+    }
+
+    /// Request the results of an in-progress collection from the leader aggregator. This may
+    /// return `Ok(None)` if the aggregation is not done yet.
+    #[tracing::instrument(err)]
+    pub async fn poll_once(
+        &self,
+        job: &CollectJob<V>,
+    ) -> Result<Option<V::AggregateResult>, Error> {
+        let response = retry_http_request(
+            self.parameters.http_request_retry_parameters.clone(),
+            || async {
+                let mut request = self.http_client.get(job.collect_job_url.clone());
+                match &self.parameters.authentication {
+                    Authentication::BearerToken(token) => {
+                        request = request.header(DAP_AUTH_HEADER, token)
+                    }
+                }
+                request.send().await
+            },
+        )
+        .await
+        .map_err(|res| match res {
+            Ok(response) => Error::Http(response.status()),
+            Err(error) => Error::HttpClient(error),
+        })?;
+
+        let status = response.status();
+        match status {
+            StatusCode::ACCEPTED => return Ok(None),
```

Comment:
IIRC the protocol text urges aggregator implementations to set a [`Retry-After`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After) header, so it'd be nice to check that here and perhaps provide it to the caller so they can decide when to try again. i.e.,
```
enum PollResult {
    AggregateResult(V::AggregateResult),
    NextAttempt(StdDuration),
}
fn poll_once(...) -> Result<PollResult, Error> { ... }
```
That said, Janus doesn't yet ever set a `Retry-After` header, so I think we could just write up an issue about this and move on. On the other hand, if we wire this up in the collector library now, then we won't have to change the public API later.

Reply:
I suspect a common way to use this type will be to call `poll_until_complete` (which calls this `poll_once` method as a subroutine) -- I think the eventual change to support `Retry-After` headers should also make `poll_until_complete` respect the `Retry-After` header, if specified.

(this would be pretty hard to do while also respecting the backoff set by the user--which is one more good reason to figure out the API sooner rather than later, IMO)